### ISSUE TITLE
Support for Font Awesome and Glyphicons by default (and perhaps other icon fonts)

### ISSUE THREAD

**Issue by marclaporte** [mpdf/mpdf#49]
mPDF 6.0 has over 85 megs (uncompressed) of data in the ttfonts directory (out of 108 Megs). This has the benefit that a large number of languages are supported out of the box.

BuiltWith reports that Font Awesome is used by 22% of websites:
http://trends.builtwith.com/widgets/fonts

Glyphicons are in Bootstrap, which has a huge presence as well:
http://trends.builtwith.com/docinfo/Twitter-Bootstrap
http://getbootstrap.com/components/#glyphicons

So they definitely should be handled somehow. And perhaps other popular open fonts should be included as well (ex.: Glyphicons)

Thanks!

**Comment by marclaporte** [comment]
Larger discussion about font handling should go here: https://github.com/mpdf/mpdf/wiki/Font-handling-in-mPDF

**Comment by mquadrat** [comment]
AFAIK mpdf does not support the :before css selector which is needed to correctly display Bootstrap icons. So some preprocessing might be needed.

**Comment by marclaporte** [comment]
Is this a lot of work?

**Comment by jonnybradley** [comment]
Looks like that's it, thanks for the clues! I managed to get some FA icons to appear in the mPDF output by using this html in the source page for the flag icon:

    <span class="fa fa-fw ">&#xf024</span>

and this in config_fonts.php font data array:

	'fontawesome' => array(
		'R' => 'fontawesome-webfont.ttf',
	),

So, we (Tiki) could do some preprocessing on our pages to render the icons like this before sending the page to "print", but a solution in mPDF would obviously be preferable so it would work for everyone.

Thanks for getting the ball rolling on this @marclaporte

**Comment by garak** [comment]
The current way of handling font is ugly. This library should avoid shipping fonts and be flexible about font location, since we cant's put stuff in `vendor` directory by hand

**Comment by finwe** [comment]
Yes, all this is planned for version 7.0 (currently unfortunately without roadplan). See https://github.com/mpdf/mpdf/blob/namespaces/src/Config/ConfigVariables.php#L410 (plus potential availability to list multiple sources).

**Comment by finwe** [comment]
Anyway, @garak, you can set custom font path in <`7.0` with `_MPDF_TTFONTPATH` constant.

**Comment by garak** [comment]
Also, setting variables with constant is ugly :cry: 
Hope to see a roadmap for 7.0 soon

**Comment by finwe** [comment]
Moved to separate issue https://github.com/mpdf/mpdf/issues/146

Roadmap for 7.0 can be seen at https://github.com/mpdf/mpdf/milestones/7.0 - with no due date, however

**Comment by jamesfairhurst** [comment]
I've managed to get Font Awesome working by using composer to pull in the package `"fortawesome/font-awesome": "*"`, defined a custom path when building the PDF:

```
define('_MPDF_SYSTEM_TTFONTS_CONFIG', FULL_PATH_TO_FILE);
$mpdf = new Mpdf();
...
```

Then included the font files in that file:

```
<?php

$this->fontdata['fontawesome'] = array(
    // relative to the mpdf fonts dir
    'R' => "/../../../fortawesome/font-awesome/fonts/fontawesome-webfont.ttf",
    'I' => "/../../../fortawesome/font-awesome/fonts/fontawesome-webfont.ttf",
);

```

Then finally in the html I can use:

```
<span class="fa fa-phone" style="font-family: fontawesome;">&#xf095;</span>
<span style="font-family: fontawesome;">&#xf095;</span>
```

Note the class isn't actually required, but useful for quickly knowing what icon it is. Hope that helps others whilst 7 is in development.

**Comment by dimvic** [comment]
something in the lines of

```php
$data = file_get_contents('path/to/fontawesome/variables.less');
$data = trim(implode("\n", array_map(function($v) { return preg_replace('/^(?:(?!fa-var).)*$/', '', $v); }, explode("\n", $data))));
$data = strtr($data, ['@fa-var'=>'fa', '\\'=>'&#x', '-'=>'_', '"'=>'']);
$data = preg_replace('/\s*:\s*/', ':', $data);
foreach (explode("\n", $data) as $line) {
    list($var, $val) = explode(':', $line);
    ${$var} = $val;
}
```

can help produce more readable templates

```php
<span style="font-family: fontawesome;"><?php echo $fa_plus; ?></span>
```

**Comment by marclaporte** [comment]
Since we'll no longer use the Github wiki, I am copying below the content from https://github.com/mpdf/mpdf/wiki/Font-handling-in-mPDF which was written by Daniel James


---
This is to brainstorm on how to improve font handling in mPDF

Related links
* [Support for Font Awesome and Glyphicons by default (and perhaps other icon fonts)](https://github.com/mpdf/mpdf/issues/49)

The mPDF font configuration needs to be more flexible so it
can take account of whichever fonts are available locally or via
@font-face. We might start this by creating configuration snippets for
popular fonts, which can be inserted into the mPDF configuration
automatically if the corresponding file names are present in the font path.

In general, mPDF should not bundle any fonts itself, as this makes
redistribution difficult (for example Debian or Ubuntu already ship some
of mPDF's fonts, but not others, and these distros don't want to ship
duplicates). It also means that the bundled fonts can get stale, which
is particularly an issue for icon fonts which are updated regularly with
new glyphs.

My suggestion would be to provide mPDF with scripts that can download
the required fonts at installation time, for example
https://packages.debian.org/jessie/fonts-font-awesome if your server
runs Debian. Most Western users do not need the CJK or Arabic fonts,
which are the biggest downloads.

**Comment by pupi** [comment]
In my project, I have to stick with version 6 for now. To work around, I use @shark0der 's answer from [here](https://stackoverflow.com/questions/17586409/adding-font-to-mpdf/41539351#41539351) but modify a bit. 
Instead of importing all library fonts, I define relative path to customised fonts from `vendor/mpdf/mpdf/ttfonts/`.
```
            $mpdf = new \mPDF('UTF-8');
            $fontdata['test'] = [
                'R' => '../../../../resources/assets/fonts/test.ttf',
                'I' => '../../../../resources/assets/fonts/test.ttf',
            ];
            foreach ($fontdata as $f => $fs) {
                $mpdf->fontdata[$f] = $fs;
                foreach (['R', 'B', 'I', 'BI'] as $style) {
                    if (isset($fs[$style]) && $fs[$style]) {
                        $mpdf->available_unifonts[] = $f . trim($style, 'R');
                    }
                }
            }
            $mpdf->default_available_fonts = $mpdf->available_unifonts;
```
I also don't need to define `_MPDF_TTFONTPATH`

**Comment by marclaporte** [comment]
A good use case for packages: https://github.com/mpdf/mpdf/issues/552
